polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -176,32 +209,147 @@ pub enum ArtifactState {
 
 /// A container of all known artifact ids and their states.
 pub struct Artifacts {
-	artifacts: HashMap<ArtifactId, ArtifactState>,
+	inner: HashMap<ArtifactId, ArtifactState>,
 }
 
 impl Artifacts {
-	/// Initialize a blank cache at the given path. This will clear everything present at the
-	/// given path, to be populated over time.
-	///
-	/// The recognized artifacts will be filled in the table and unrecognized will be removed.
-	pub async fn new(cache_path: &Path) -> Self {
-		// First delete the entire cache. This includes artifacts and any leftover worker dirs (see
-		// [`WorkerDir`]). Nodes are long-running so this should populate shortly.
-		let _ = tokio::fs::remove_dir_all(cache_path).await;
+	#[cfg(test)]
+	pub(crate) fn empty() -> Self {
+		Self { inner: HashMap::new() }
+	}
+
+	#[cfg(test)]
+	pub(crate) fn len(&self) -> usize {
+		self.inner.len()
+	}
+
+	/// Create an empty table and populate it with valid artifacts as [`ArtifactState::Prepared`],
+	/// if any. The existing caches will be checked by their file name to determine whether they are
+	/// valid, e.g., matching the current node version. The ones deemed invalid will be pruned.
+	pub async fn new_and_prune(cache_path: &Path) -> Self {
+		let mut artifacts = Self { inner: HashMap::new() };
+		artifacts.insert_and_prune(cache_path).await;
+		artifacts
+	}
+
+	async fn insert_and_prune(&mut self, cache_path: &Path) {
+		async fn is_corrupted(path: &Path) -> bool {
+			let file_name = path
+				.file_stem()
+				.expect("path must contain a file name")
+				.to_str()
+				.expect("file name must be valid UTF-8");
+			let checksum = match tokio::fs::read(path).await {
+				Ok(bytes) => blake3::hash(&bytes),
+				Err(err) => {
+					// just remove the file if we cannot read it
+					gum::warn!(
+						target: LOG_TARGET,
+						?err,
+						"unable to read file {:?} when checking integrity",
+						path,
+					);
+					return true
+				},
+			};
+			!file_name.ends_with(checksum.to_hex().as_str())
+		}
+
+		// Insert the entry into the artifacts table if it is valid.
+		// Otherwise, prune it.
+		async fn insert_or_prune(
+			artifacts: &mut Artifacts,
+			entry: &tokio::fs::DirEntry,
+			cache_path: &Path,
+		) {
+			let file_type = entry.file_type().await;
+			let file_name = entry.file_name();
+
+			match file_type {
+				Ok(file_type) =>
+					if !file_type.is_file() {
+						return
+					},
+				Err(err) => {
+					gum::warn!(
```

Comment:
nit: There are three warn logs in this fn ("discarding invalid artifact" should be a warn). We could reduce some LoC by returning an error and logging in a single place, at the call site.

Reply:
gonna punt on this, coz only one of them is `Err`